### PR TITLE
Add support for recursively creating child directories

### LINKED ISSUE #61: Add support for Putting Objects of foo/bar
Rather than getting issues where `ParentNotDirectory` errors, we should add the ability to allow the user to specify a path as follows:

```
/account_name/stor/test-bucket/foo/bar
```

where foo is a directory that doesn't exist. The library should handle that logic. It looks similar as to the following:

```
objectPath := path.Join(mantaDefaultRootStore, m.directory)
	parts := strings.Split(entry.Key, "/")

	//This will remove the last part only, this way, I don't have to worry
	//about checking to see if the part is actually the last part
	parentParts := parts[:len(parts)-1] //foo
	keyPart := parts[len(parts)-1]      //bar

	for _, part := range parentParts {
		objectPath = path.Join(objectPath, part) //stor/test-bucket/foo
		err := m.client.Dir().Put(context.Background(), &storage.PutDirectoryInput{
			DirectoryName: objectPath,
		})
		if err != nil {
			return err
		}
	}
	return m.client.Objects().Put(context.Background(), &storage.PutObjectInput{
		ObjectPath:    path.Join(objectPath, keyPart),
		ObjectReader:  r,
		ContentLength: uint64(len(entry.Value)),
	})
```

### PR DESCRIPTION
**Pull request by stack72**
Fixes #61

when we want to Put an object in /stor/foo/bar and foo/bar doesn't exist, we will now create that object structure

```
% ./forcePut
2017/12/15 17:04:44 Folder Path "/stor/folder1/folder2/folder3/" not found
2017/12/15 17:04:44 Forcing the Creation of the Directory Tree: "/stor/folder1/folder2/folder3/foo.txt"
Successfully uploaded /tmp/foo.txt to /stor/folder1/folder2/folder3/foo.txt
```

### CODE REVIEW

File: examples/storage/forcePut.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		log.Fatalf("NewClient: %s", err)
+	}
+
+	reader, err := os.Open("/tmp/foo.txt")
+	if err != nil {
+		log.Fatalf("os.Open: %s", err)
+	}
+	defer reader.Close()
+
+	err = client.Objects().Put(context.Background(), &storage.PutObjectInput{
```
Comment (cheapRoc) [inline examples/storage/forcePut.go]: Do we want a `ForceInsert` or something? Without it being optional we would wipe over whatever advantages we may have by reporting sub-directories as not existing in today's usage of Manta.

Thoughts?

File: storage/objects.go [hunk 2]
```diff
@@ -258,6 +260,21 @@ type PutObjectInput struct {
 func (s *ObjectsClient) Put(ctx context.Context, input *PutObjectInput) error {
 	fullPath := path.Clean(path.Join("/", s.client.AccountName, input.ObjectPath))
 
+	err := checkDirectoryTreeExists(*s, ctx, input.ObjectPath)
+	if err != nil {
+		if strings.Contains(err.Error(), "ResourceNotFound") { //ResourceNotFound when dir.List doesn't exist
```
Comment (cheapRoc) [inline storage/objects.go]: I don't normally like this given a type system. We discussed error handling yesterday though so I'm fine with this for now.

File: storage/objects.go [hunk 3]
```diff
@@ -304,3 +321,48 @@ func (s *ObjectsClient) Put(ctx context.Context, input *PutObjectInput) error {
...
+}
+
+func checkDirectoryTreeExists(c ObjectsClient, ctx context.Context, fullPath string) error {
+	dirClient := &DirectoryClient{
+		client: c.client,
+	}
+
+	folders, _ := path.Split(fullPath)
+
+	_, err := dirClient.List(ctx, &ListDirectoryInput{
```
Comment (cheapRoc) [inline storage/objects.go]: Any chance of using our new HEAD functions, `Objects.GetInfo` or `Objects.IsDir` here?

File: storage/objects.go [hunk 4]
```diff
@@ -258,6 +260,21 @@ type PutObjectInput struct {
 func (s *ObjectsClient) Put(ctx context.Context, input *PutObjectInput) error {
 	fullPath := path.Clean(path.Join("/", s.client.AccountName, input.ObjectPath))
 
+	err := checkDirectoryTreeExists(*s, ctx, input.ObjectPath)
```
Comment (cheapRoc) [inline storage/objects.go]: If we're not checking within the loop how are we handling existing directories?

File: storage/objects.go [hunk 5]
```diff
@@ -304,3 +321,48 @@ func (s *ObjectsClient) Put(ctx context.Context, input *PutObjectInput) error {
...
+		if i == 0 {
+			folderPath = part
+			continue
+		}
+		folderPath = path.Clean(path.Join(folderPath, part))
+		err := dirClient.Put(ctx, &PutDirectoryInput{
+			DirectoryName: folderPath,
+		})
+		if err != nil {
+			return err
```
Comment (cheapRoc) [inline storage/objects.go]: Will this error out if we can't create a directory along the way for whatever reason?

File: examples/storage/forcePut.go (hunk 1 above)
Comment (stack72) [reply examples/storage/forcePut.go]: Love this thought - much safer!

File: storage/objects.go (hunk 2 above)
Comment (stack72) [reply storage/objects.go]: Yeah error handling needs to change all over the shop!

File: storage/objects.go (hunk 5 above)
Comment (stack72) [reply storage/objects.go]: yes, it will error out and report that error to the user

File: storage/objects.go (hunk 3 above)
Comment (stack72) [reply storage/objects.go]: This is a check to see if the folder path exists - not if it's a directory or an object

File: examples/storage/forcePut.go (hunk 1 above)
Comment (stack72) [reply examples/storage/forcePut.go]: Added :)

File: storage/objects.go (hunk 4 above)
Comment (stack72) [reply storage/objects.go]: Demo'ed to @cheapRoc that we have a test case for this to work and works as expected

File: storage/objects.go (hunk 2 above)
Comment (stack72) [reply storage/objects.go]: We have removed at least some of the string matching for now

File: storage/objects.go (hunk 3 above)
Comment (stack72) [reply storage/objects.go]: Changed to using IsDir :)

**Pull request merged**